### ISSUE TITLE
Connections from one Android device to Another

### ISSUE THREAD

**Issue by insominx** [don/BluetoothSerial#50]
I know this was intended to connect with Arduino but it seems like connecting between a pair of android or ios devices could work too.  Is this already the case?  I fiddled with the examples for a while (using a nexus 7 and galaxy s4) but could not establish a connection.

The examples don't appear to have code to receive a connection in the javascript.  Is that only in the arduino logic?

**Comment by tanzeelrana** [comment]
This version does not work for me. 
- I downloaded the project
- copied the Chat example in a separate place
- cordova plugin add com.megster.cordova.bluetoothserial
- cordova platform add android
- cordova build - > and run 

but again I see the same issue. when I try to connect my nexus 4 to my nexus 5  or vice versa. it instantly says disconnected. Was anyone successful in connecting 2 android devices with https://github.com/tomvanenckevort/BluetoothSerial  ?

**Comment by ragcsalo** [comment]
I have updated the Java and JavaScript files, these are working just fine.
:-)
Please feel free to use them!
Regards, Bence

2015-02-19 18:49 GMT+01:00 tanzeelrana <notifications@github.com>:

**Comment by ragcsalo** [comment]
Download updated files from here:
http://download.webtanar.hu/BluetoothSerial/updated_files.zip

**Comment by tanzeelrana** [comment]
thanks !! @ragcsalo  I am going to try this out ..... 

https://github.com/tomvanenckevort/BluetoothSerial

have you tried this app? my last comment was regarding this link above. I tried this and it was not connecting 2 android devices

**Comment by ragcsalo** [comment]
No, haven't tried that one. This one I updated myself, and have been using
it to connect Samsung devices for 8 months now... :-)

**Comment by tanzeelrana** [comment]
Under limitations it says 

Limitations

The phone must initiate the Bluetooth connection
iOS Bluetooth Low Energy requires iPhone 4S, iPhone5, iPod 5, or iPad3+
Will not connect Android to Android*
Will not connect iOS to iOS*

does that need to be changed ?

**Comment by ragcsalo** [comment]
With my modified source files you CAN connect Android to Android. I don't
know how to create a forked version, but will try it someday. :-)

**Comment by tanzeelrana** [comment]
@ragcsalo  with your updated files the app is forever on discovering devices
or it just normally takes long time to discover devices ?

**Comment by ragcsalo** [comment]
I don't discover devices, I just display a dropdown list of paired devices,
select one and connect. On the second device I also have to select and
connect the first device.... and then can send and receive data. Shall I
make a sample file?

**Comment by tanzeelrana** [comment]
If you can zip the project that works for you and share I will try to find the differences in code and figure out what is causing this problem.

**Comment by ragcsalo** [comment]
I'll upload the project ZIP tomorrow. :-)

**Comment by tanzeelrana** [comment]
Thanks :+1:

**Comment by tanzeelrana** [comment]
@ragcsalo  did you upload the project ?

**Comment by ragcsalo** [comment]
Sorry for the delay. I have updated the source files from Hungarian to
English... :-)
You can download the project here:
http://download.webtanar.hu/BluetoothSerial/bluetooth.zip

1) pair the 2 Android devices over Bluetooth first
2) install bluetooth.apk on both devices
3) after starting the app, select other device from dropdown list (on both
devices)
4) click on "Connect" on both devices
5) after connected, you can send messages over Bluetooth :)

**Comment by Slim2k** [comment]
Hi, 

I know this is 5 months old since last activity, but I am trying to send data between 2 Android devices (Nexus 7 and Moto X 2nd Gen) I Can get them to connect and stay connected using the .connect() and  isConnected() functions. 

But when I use the Write() function, I get a success callback that the string was sent, but neither read() or Subscribe is receiving the data. Really not sure what else to try, code is very messy as I have been trying all different things, but below.,

    //Hosting the game
    $scope.isHost = null;
    //Watching the game
    $scope.isGuest = null;

    var Delimiter = '\n';
    var DeviceAddress;

    var isTurnedOn
    var isEnabled;
    var isConnected;
    var sender;
    var reciver;

    //Check is Bluetooth is turned on
    $scope.btEnabled = function () {
        $cordovaBluetoothSerial.isEnabled().then(
                function () {
                    isEnabled = "Bluetooth is enabled";
                },
                function () {
                    btEnabled = "Bluetooth is NOT enabled";
                }
        )
        return isTurnedOn;
    };

    //Check if device is connected
    var s = ""
    $scope.connectionStatus = function () {
        setTimeout(function () {
            $cordovaBluetoothSerial.isConnected().then(
                    function (info) {
                        console.log("BT Connected");
                        s = "Bluetooth is connected";
                        isConnected = true;
                    },
            function () {
                console.log("BT *not* connected");
                s = "Bluetooth is not connected";
                isConnected = false;
            });
        }, 5000);

        return "is "+s;
    };

    //List of BT devices
    $scope.btList = {}
    $scope.btDiscover = function () {
        $cordovaBluetoothSerial.list().then(
            function (devices) {
                $scope.btList = devices;
                $scope.showBTlist = true;
            },
            function (err) {
                console.log(err);
                $scope.showBTlist = false;
            }
        )
    };
    
    function subScribe() {
        $cordovaBluetoothSerial.subscribe('\n',
            function (data) {
                console.log(data);
                $scope.subscribData = "Sub Data = " + data;
        });
    }


    $scope.btConnect = function (item) {

        //Determin devce
        if (item.address) {
            //Adress Android and WP
            DeviceAddress = item.address;
        }
        else if (item.uuid) {
            // UUID used by iOS
            DeviceAddress = item.uuid;
        }
        
        if (isConnected != true) {
            sender = true;
            $cordovaBluetoothSerial.connect(DeviceAddress).then(
                function (success) {
                    console.log('Connected');
                    $scope.connectedTo = item.name;
                    // subScribe();
                    //stayConnected();
                },
                function (err) {
                    console.log('Connection failed');
                    $scope.connectedTo = item.name + err;
                }
            )
        }
        else {
            console.log("Subscribe");
            reciver = true;
            subScribe();
        }
    };


    $scope.sendMessage = function (message) {
        //$cordovaBluetoothSerial.write('\n' + message);
        $cordovaBluetoothSerial.write('\n' + message).then(
            function (success) {
                console.log("Message Sent")
            },
            function (failure) {
                console.log("Message Sent" + failure)
            }
        )
    }

**Comment by mitosistech** [comment]
Can you please help me ?I have tried this bluetooth serial plugin.But i can't able to connect the two android device.Please suggest ,what I need to change in code.I am new to cordova

**Comment by BurakDev** [comment]
@don actually support android-to-android ? ios-to-ios ? android-to-ios ?

**Comment by ateufel** [comment]
@BurakDev there is no support for that, but you can just take a look at the bluetooth chat example in the official android docs for android to android communication.

**Comment by ragcsalo** [comment]
I already uploaded my code, I made Android to Android connection with a
modified version of Don's bluetooth serial plugin.
2015. dec. 9. 0:11 ezt írta ("Andreas Teufel" <notifications@github.com>):

**Comment by ateufel** [comment]
@ragcsalo the link to the zip file does not work anymore

**Comment by don** [comment]
@BurakDev There's a fork that should support Android to Android https://github.com/don/BluetoothSerial/issues/50#issuecomment-66405396

The original code for Android was based on the Android chat example that @ateufel mentioned. I think people enabled the RFCOMM listening to make this work.

Maybe someone can issue a pull request based on @tomvanenckevort or @ragcsalo's changes?

For iOS to iOS, you'll need one of the iOS devices to be running a Bluetooth Low Energy service that looks like one of the supported UART services. (Nordic/Adafruit, RedBearLab, BlueGiga, Laird)

For Android to iOS or iOS to Android, you'd need Bluetooth Low Energy, since iOS doesn't (easily) support Bluetooth Classic. You can use another one of my plugins https://github.com/don/cordova-plugin-ble-central to connect to Bluetooth low energy services. You'd need to create the service with native code for now. (I have cordova-plugin-ble-peripheral planned, but not implemented).

**Comment by ragcsalo** [comment]
Uhh... I accidentaly deleted the ZIP file... I'll upload it again soon!!!

**Comment by cjtoribio** [comment]
@ragcsalo I am new with this library and saw all the comments on connecting two android devices as i tried and failed. I see the ZIP files are not up. I could help you on forking the solution.

**Comment by ateufel** [comment]
@cjtoribio just take a look at the official bluetooth chat example, it´s a pretty good tutorial with all the code your need: https://developer.android.com/samples/BluetoothChat/index.html

**Comment by gcdpinho** [comment]
I am having a problem to send and receive messages between two devices android. I place the connection and the function write returns me sucess , but I can't receive the message on another device. 

Srry abount my english, i'm brazilian.

**Comment by ateufel** [comment]
that´s a bit too broad, i´m afraid. all we know is "it doesn't work", which is nothing we can work with. take a look at the links in my comment above, it should be quite easy to figure out what´s wrong with the original code from google.

**Comment by paredesivan** [comment]
@ragcsalo could you upload the file again?

**Comment by sando431** [comment]
@ragcsalo 

http://download.webtanar.hu/BluetoothSerial/bluetooth.zip

I am seeing 404 error on this page .

Also i tried your changes and in both android devices it is just showing discovering message in drop-down title and nothing is happening.

Can you please help me regarding this.

**Comment by ragcsalo** [comment]
I have uploaded the files, instructions and the whole project again! :-)

Download updated files from here (instructions and final .APK of my test app included):
http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip

Download the full Android Studio project from here:
http://benkesmith.com/android/bt_serial/btserial_project.zip

**Comment by ragcsalo** [comment]
# Instructions

```
1) Create your Cordova project as usual:
========================================
cordova create project_name com.yourdomain.project_name project_name
cd project_name
cordova platform add android
cordova build android

2) Add the necessary plugins to your project (select the ones you need):
=============================================
cordova plugin add cordova-plugin-inappbrowser
cordova plugin add cordova-plugin-device
cordova plugin add cordova-plugin-dialogs


3) Add the BluetoothSerial plugin:
==================================
cordova plugin add https://github.com/don/BluetoothSerial.git


4) Copy the working .js and .java files from this .zip file (to your project, overwriting the old files):
===========================================================
platforms/android/app/src/main/assets/www/plugins/cordova-plugin-bluetooth-serial/www/bluetoothSerial.js
platforms/android/app/src/main/java/com/megster/cordova/BluetoothSerial.java
platforms/android/app/src/main/java/com/megster/cordova/BluetoothSerialService.java


5) In your main file (index.html) make the BT connection:
=========================================================
See the HTML and JavaScript code in platforms/android/app/src/main/assets/www/index.html
The necessary js/jquery-1.9.1.min.js is also included in this .zip file.


6) Build the project, and install the .apk on both Android devices


7) Turn on Bluetooth and pair your devices (on the Bluetooth settings page of your devices)


8) Run the app, and click on 'Connect BT remote' on BOTH devices


9) When the connection is OK, send a message to the other device
   Type a message in the text field, and press 'Send message' button.

```

**Comment by sando431** [comment]
Thanks a lot ragcsalo .It will really help me.I just installed your apk and its working as expected.Will go through the code now and ask you if any help needed in understanding..

Thanks again.... :)

**Comment by htahir6200** [comment]
Hello  @ragcsalo   , Hope you are good and enjoying,  very nice application this one. I have use it for my android to android connection.  But i want to use the same app for webrtc signalling. But i dont have any idea how to do that , could you please help me in  doing this.

**Comment by flockf** [comment]
@tanzeelrana Did you manage to get this one to work: https://github.com/tomvanenckevort/BluetoothSerial

or anyone else?

When i click connect on the 2nd device, it disconnects from the 1st device (meaning they were connected when i clicked connect on the 1st device, but no messages could get sent). 

Anyone has a solution for this?

**Comment by wang3g** [comment]
See the same problem here as @flockf mentioned

**Comment by ragcsalo** [comment]
Have you tried my ZIP file and instructions? :-)

**Comment by wang3g** [comment]
Yes, I downloaded the ZIP file from http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip, and followed the exact instructions given on on Oct 6, 2016. 

One device I m using is Huawei Mate 9 phone, the other one is Samsung tablet.

My trial project is an ionic 1.x project though.

I

**Comment by ragcsalo** [comment]
Hmmm.... I don't know what could cause the connection error. If others have the same problem, it could be the plugin code that's buggy. If only you have the problem, it must be something wrong in your code... :-/

**Comment by flockf** [comment]
@ragcsalo i got yours to work man, it is just a bit confusing

**Comment by ragcsalo** [comment]
Great! :-) I was also experimenting with it, changing a code a little, and now it works well most of the time. :-)

**Comment by flockf** [comment]
@ragcsalo Thanks man, you did an awesome job. And thanks for commenting

**Comment by ragcsalo** [comment]
My pleasure to help others :-)

**Comment by wang3g** [comment]
@flockf, is it possible for you to share what you have done to fix the connection problem? Thanks a lot.

**Comment by flockf** [comment]
@wang3g  Well i just tried again, carefully following the instructions. I think i had missed adding the plugins which the example of @ragcsalo uses.
Have you tested the demo apk provided by @ragcsalo ?

**Comment by wang3g** [comment]
@flockf Yes, the demo apk provided by @ragcsalo works. I also tried again and followed the instructions from step 1 to 9, but still no luck

**Comment by flockf** [comment]
@wang3g If you are simply trying to build the chat demo provided, what does your project's folder structure look like? There must be something you are missing. Did you add the three other plugins? 
What problem do you have exactly?

**Comment by wang3g** [comment]
![screen shot 2017-08-23 at 12 42 08 pm](https://user-images.githubusercontent.com/6946817/29596441-84f4a236-8800-11e7-8293-9acbd9ed7a39.png)

@flockf  Above is my project's folder structure.
I have added all 3 plugins:
cordova plugin add cordova-plugin-inappbrowser
cordova plugin add cordova-plugin-device
cordova plugin add cordova-plugin-dialogs

The problem I m seeing here is:
When i click connect on the 2nd device, it disconnects from the 1st device. And when i click connect on the 1st device, it disconnects from the 2nd device.

**Comment by flockf** [comment]
@wang3g if you have not figured it out yet, feel free to upload your project and i will have a look !

**Comment by flockf** [comment]
@ragcsalo Man, is it possible to handle multiple, simultaneous connections with yours?

**Comment by ragcsalo** [comment]
@flockf I don't think so... :-/

**Comment by flockf** [comment]
@ragcsalo ohhh. would this work? what do you think?

**Comment by I-Connect** [comment]
Hi, I have the same issue as der1typ. Seems package.json is missing which prevents the jackTheRipper/BluetoothSerial from installing. Any work around known for this?

thx!

**Comment by turkalpjr** [comment]
@ragcsalo   
Hello, 
I deeply searched web, tried to accomplish but am at the place,where I began, no progress. your solution 

"
I have uploaded the files, instructions and the whole project again! :-)

Download updated files from here (instructions and final .APK of my test app included):
http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip

Download the full Android Studio project from here:
http://benkesmith.com/android/bt_serial/btserial_project.zip
"
which is, what I exactly try to accomplish in ionic 2 , that you done in pure android.
I think that you deserve to be put on ionic page something like 

Repo: https://github.com/ragscalo/BluetoothSerial  

Anyway, My aim is , using ionic2    bluetooth-serial or  ionic2    bluetooth- ble  , I want to send hello world message from one android phone to another android phone.

for this purpose, I prepared an ionic 3 app...

http://blog.turkalpkucur.com/debug.rar

http://blog.turkalpkucur.com/error.jpg

**Comment by Alexander055** [comment]
@ragcsalo 

Thank for your instruction, it works for me,

Just a little bug I found, once the two device are connected, if I try to press disconnect on the device I start the connection, and so press disconnect on the other device, I am not able to reconnect them again by pressing connecting on the first device.


1) Device_02 I press connect BT remote 
2) Device_01 I press connect BT remote 
3) now the two devices are connected each other
4) Press disconnect on Device_01
5) Press disconnect on Device_02

Now Device_01 is unabled to reconnect to Device_02 by pressing Connect BT remote

**Comment by Paras-avantsoft** [comment]
@ragcsalo Thanks man for your solution. I downloaded from below link and the apk works great with my android devices. I am able to transfer the msg from one device to other.
- http://benkesmith.com/android/bt_serial/btserial_project.zip

But when I tried to do this with latest version 0.4.7 this is not working. Its always showing me that 'Unable to connect the device'. I also check the documentation of @dom, and its saying that android to android connection will not work.

Is there any solution you have for this?

**Comment by ragcsalo** [comment]
I always copy the .java and .js files from one of my earlier projects after I install the plugin.
I should've made a fork or a complete copy of the plugin when everything was working...
Lately I've been experiencing with BT LE plugins, and I could easily make connection between 2 Android (and also iOS) devices! One is the master (central), one is the slave (peripherial), and they both can send/receive data. :-) Check out this plugin: https://github.com/randdusing/cordova-plugin-bluetoothle

**Comment by Paras-avantsoft** [comment]
@ragcsalo 

Thanks for the quick reply.
I will try by replacing the .java and .js files and will also try your suggested plugin and let you know if its working for me.

**Comment by Paras-avantsoft** [comment]
@ragcsalo

Replacing the .java and .js file worked. Thanks man for the help.

**Comment by turkalpjr** [comment]
hey fellas,
thanks ragscalo, but I have to mention that, platform independent solution, native solution (independent from being ios or andro) must be served. thus, the solution of don is a crap! it is not working! I hope, in ionic 4, these issues will be solved. so , there is no solution.

**Comment by w4ugit** [comment]
Hello. I try to make the application using the instruction and I do not go out, the devices do not connect. can you help me? I really need

**Comment by ragcsalo** [comment]
Hello guys,
I have updated the Instructions (see above, and also in the Bluetooth_Serial.zip file) and also the ZIP files. :-) If you've failed to connect 2 Android devices, try again with the new files:

Download updated files from here (updated instructions and final .APK of my test app included):
http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip

Download the full Android Studio v3.0 project from here:
http://benkesmith.com/android/bt_serial/btserial_project.zip

Good luck! :-)

**Comment by flockf** [comment]
@ragcsalo 
Hey man, is it possible to get subscribeRaw to work on your version of the plugin? 
How?

Thanks!

**Comment by longas** [comment]
@ragcsalo thanks for your work.

I've used your solution in a Ionic (v4) project and this is what I had to do to make it work.

- First add the plugin to your project:
`ionic cordova plugin add cordova-plugin-bluetooth-serial`
`npm install @ionic-native/bluetooth-serial`

- [Download @ragcsalo solution](http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip)

- Go to _platforms\android\app\src\main\assets\www\plugins\cordova-plugin-bluetooth-serial\www_ , open bluetoothSerial.js and remove the first line:
`cordova.define("cordova-plugin-bluetooth-serial.bluetoothSerial", function(require, exports, module) {});`

- Then copy that file to your project's _plugins\cordova-plugin-bluetooth-serial\www_

- Copy _platforms\android\app\src\main\java\com_ from ragcsalo's package to your project's _plugins\cordova-plugin-bluetooth-serial\src\android_

- I recommend removing your android build located in _platforms\android_ and then running:
`ionic cordova build android`

**Comment by sammy00747** [comment]
> @ragcsalo thanks for your work.
> 
> I've used your solution in a Ionic (v4) project and this is what I had to do to make it work.
> 
> * First add the plugin to your project:
>   `ionic cordova plugin add cordova-plugin-bluetooth-serial`
>   `npm install @ionic-native/bluetooth-serial`
> * [Download @ragcsalo solution](http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip)
> * Go to _platforms\android\app\src\main\assets\www\plugins\cordova-plugin-bluetooth-serial\www_ , open bluetoothSerial.js and remove the first line:
>   `cordova.define("cordova-plugin-bluetooth-serial.bluetoothSerial", function(require, exports, module) {});`
> * Then copy that file to your project's _plugins\cordova-plugin-bluetooth-serial\www_
> * Copy _platforms\android\app\src\main\java\com_ from ragcsalo's package to your project's _plugins\cordova-plugin-bluetooth-serial\src\android_
> * I recommend removing your android build located in _platforms\android_ and then running:
>   `ionic cordova build android`

@longas it says "plugin not installed" as a warning

**Comment by sunilk44** [comment]
@ragcsalo i have tried your code . every thing is working except  bluetoothSerial.subscribe function which is responsible for recieving data to other end. subscribe function not responding with failur or success. I followed all your steps but didn't abel to get why subscribe function is not running.
could you suggest me some solution for that or u got that issue while make this project.

**Comment by akashkhairnar01** [comment]
@ragcsalo i follow all the instruction but it is not working for me

**Comment by 0xclx** [comment]
I am asking here because I have no other way of finding out about my Problem.

Does anyone else of you using this plugin have the problem of sending (writing) data to a remote device (SPP Connection, NOT Android to Android) where messages get lost/replies are somtimes not received?

**Comment by chtenb** [comment]
Is this still the most up-to-date solution and does it work for capacitor?

> I've used your solution in a Ionic (v4) project and this is what I had to do to make it work.
> 
> * First add the plugin to your project:
>   `ionic cordova plugin add cordova-plugin-bluetooth-serial`
>   `npm install @ionic-native/bluetooth-serial`
> * [Download @ragcsalo solution](http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip)
> * Go to _platforms\android\app\src\main\assets\www\plugins\cordova-plugin-bluetooth-serial\www_ , open bluetoothSerial.js and remove the first line:
>   `cordova.define("cordova-plugin-bluetooth-serial.bluetoothSerial", function(require, exports, module) {});`
> * Then copy that file to your project's _plugins\cordova-plugin-bluetooth-serial\www_
> * Copy _platforms\android\app\src\main\java\com_ from ragcsalo's package to your project's _plugins\cordova-plugin-bluetooth-serial\src\android_
> * I recommend removing your android build located in _platforms\android_ and then running:
>   `ionic cordova build android`

**Comment by bevinGithub** [comment]
Hi guys is this solution also applicable on Ionic 6  targeting SDK 31 Android 1, l keep getting  error below

"Need android.permission.BLUETOOTH_SCAN permission for AttributionSource { uid = 10467, packageName = com.fire.service.app, attributionTag = null, token = android.os.BinderProxy@a208f16, next = null }: Starting discovery."

Your assistance will be much appreciated.

Below is my environment
Ionic:
   Ionic CLI                     : 6.20.1 
   Ionic Framework               : @ionic/angular 6.2.2
   @angular-devkit/build-angular : 13.0.4
   @angular-devkit/schematics    : 13.0.4
   @angular/cli                  : 13.0.4
   @ionic/angular-toolkit        : 5.0.3

Cordova:
   Cordova CLI       : 11.0.0
   Cordova Platforms : android 10.1.2
   Cordova Plugins   : cordova-plugin-ionic-keyboard 2.2.0, cordova-plugin-ionic-webview 5.0.0, (and 14 other plugins)

Utility:
   cordova-res                          : 0.15.4
   native-run (update available: 1.7.1) : 1.5.0

System:
   NodeJS : v14.15.4 (C:\Program Files\nodejs\node.exe)
   npm    : 6.14.10
   OS     : Windows 10

**Comment by ash737** [comment]
> > @ragcsalo thanks for your work.
> > I've used your solution in a Ionic (v4) project and this is what I had to do to make it work.
> > 
> > * First add the plugin to your project:
> >   `ionic cordova plugin add cordova-plugin-bluetooth-serial`
> >   `npm install @ionic-native/bluetooth-serial`
> > * [Download @ragcsalo solution](http://benkesmith.com/android/bt_serial/Bluetooth_Serial.zip)
> > * Go to _platforms\android\app\src\main\assets\www\plugins\cordova-plugin-bluetooth-serial\www_ , open bluetoothSerial.js and remove the first line:
> >   `cordova.define("cordova-plugin-bluetooth-serial.bluetoothSerial", function(require, exports, module) {});`
> > * Then copy that file to your project's _plugins\cordova-plugin-bluetooth-serial\www_
> > * Copy _platforms\android\app\src\main\java\com_ from ragcsalo's package to your project's _plugins\cordova-plugin-bluetooth-serial\src\android_
> > * I recommend removing your android build located in _platforms\android_ and then running:
> >   `ionic cordova build android`
> 
> @longas it says "plugin not installed" as a warning

Hi @sammy00747,
I have same issue ` plugin not installed` 
Did you got a solution?
